pylabrobot/plate_reading/biotek_backend.py

```diff
@@ -0,0 +1,250 @@
+    resp = await self.send_command("C")
+    assert resp is not None
+    return resp[1:].split(b" ")[0].decode()
+
+  async def get_firmware_version(self) -> str:
+    resp = await self.send_command("e")
+    assert resp is not None
+    return " ".join(resp[1:-1].decode().split(" ")[0:4])
+
+  async def open(self):
+    return await self.send_command("J")
+
+  async def close(self):
+    return await self.send_command("A")
+
+  async def get_current_temperature(self) -> float:
+    """ Get current temperature in degrees Celsius. """
+    resp = await self.send_command("h")
+    assert resp is not None
+    return int(resp[1:-1]) / 100000
+
+  def _parse_body(self, body: bytes) -> List[List[float]]:
+    start_index = body.index(b"01,01")
+    end_index = body.rindex(b"\r\n")
+    num_rows = 8
+    rows = body[start_index:end_index].split(b"\r\n,")[:num_rows]
+
+    parsed_data: List[List[float]] = []
+    for row_idx, row in enumerate(rows):
+      parsed_data.append([])
+      # (well, row, value)
+      values = row.split(b",")
+      grouped_values = [values[i:i+3] for i in range(0, len(values), 3)]
+
+      for group in grouped_values:
+        assert len(group) == 3
+        value = float(group[2].decode())
+        parsed_data[row_idx].append(value)
+    return parsed_data
+
+  async def read_absorbance(self, wavelength: int) -> List[List[float]]:
+    if not 230 <= wavelength <= 999:
+      raise ValueError("Wavelength must be between 230 and 999")
+
+    resp = await self.send_command("y", wait_for_char=b"\x06")
+    assert resp == b"\x06"
+    await self.send_command(b"08120112207434014351135308559127881772\x03", purge=False)
+
+    resp = await self.send_command("D", wait_for_char=b"\x06")
+    assert resp == b"\x06"
+    wavelength_str = str(wavelength).zfill(4)
+    cmd = f"00470101010812000120010000110010000010600008{wavelength_str}1".encode()
+    checksum = str(sum(cmd) % 100).encode()
+    cmd = cmd + checksum + b"\x03"
+    await self.send_command(cmd, purge=False)
+
+    resp1 = await self.send_command("O", wait_for_char=b"\x06")
+    assert resp1 == b"\x06"
+    resp2 = await self._read_until(b"\x03")
+    assert resp2 == b"0000\x03"
+
+    # read data
+    body = await self._read_until(b"\x03")
+    assert resp is not None
+    return self._parse_body(body)
+
+  async def read_luminescence(self, focal_height: float) -> List[List[float]]:
+    raise NotImplementedError("Not implemented yet")
+
+  async def read_fluorescence(
+    self,
+    excitation_wavelength: int,
+    emission_wavelength: int,
+    focal_height: float,
+  ) -> List[List[float]]:
+    if not 4.5 <= focal_height <= 13.88:
+      raise ValueError("Focal height must be between 4.5 and 13.88")
+    if not 250 <= excitation_wavelength <= 700:
+      raise ValueError("Excitation wavelength must be between 250 and 700")
+    if not 250 <= emission_wavelength <= 700:
+      raise ValueError("Emission wavelength must be between 250 and 700")
```

Comment:
have you considered using `pydantic` for validation? it's a nice library

Reply:
i am generally opposed to adding dependencies. in my opinion, these checks are easy enough to write and read